docs/root/intro/arch_overview/http/http_routing.rst

```diff
@@ -192,3 +192,63 @@ upon configuration load and cache the contents.
 
 If **response_headers_to_add** has been set for the Route or the enclosing Virtual Host,
 Envoy will include the specified headers in the direct HTTP response.
+
+Routing Via Generic Matching
+----------------------------
+
+Envoy recently added support for utilzing a :ref:`generic match tree <arch_overview_matching_api>` to
+specify the route table. This is a more expressive matching engine than the original one, allowing
+for sublinear matching on arbitrary headers (unlike the original matching engine which could only
+do this for :authority in some cases).
+
+To use the generic matching tree, specify a matcher on a virtual host with a RouteAction action:
+
+.. code-block:: yaml
+
+  matcher:
+    "@type": type.googleapis.com/xds.type.matcher.v3.Matcher
+    matcher_tree:
+      input:
+        name: request-headers
+        typed_config:
+          "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+          header_name: :path
+      exact_match_map:
+        map:
+          "/new_endpoint/foo":
+            action:
+              name: route
+              typed_config:
+                "@type": type.googleapis.com/envoy.config.route.v3.Route
+                match:
+                  prefix: /
+                route:
+                  cluster: cluster_foo
+                request_headers_to_add:
+                - header:
+                    key: x-route-header
+                    value: new-value
+          "/new_endpoint/bar":
+            action:
+              name: route
+              typed_config:
+                "@type": type.googleapis.com/envoy.config.route.v3.Route
+                match:
+                  prefix: /
+                route:
+                  cluster: cluster_bar
+                request_headers_to_add:
+                - header:
+                    key: x-route-header
+                    value: new-value
+
+This allows resolving the same Route proto message used for the `routes`-based routing using the additional
+matching flexibility provided by the generic matching framework.
+
+Note that the resulting Route also specifies a match criteria. This must be satisfied in addition to resolving
+the route in order to achieve a route match. When path rewrites are used, the matched path will only depend on
+the match criteria of the resolved Route. Path matching done during the match tree traversal does not contribute
+to path rewrites.
+
+The only inputs supported are request headers (via `envoy.type.matcher.v3.HttpRequestHeaderMatchInput`). See
```

Comment:
Where is this validated and what error message does the user get if they configure something else accidentally?

Same question about the action config which I guess will only currently work for route proto lookup?

Reply:
Seems like this was not validated, now we are with the following error message:

```
requirement violation while creating route match tree: INVALID_ARGUMENT: Route table can only match on request headers, saw type.googleapis.com/envoy.type.matcher.v3.HttpResponseHeaderMatchInput
```

For action configs the only registered factory for `ActionFactory<RouteActionContext>` is `RouteMatchActionFactory`, so all others would result in the typical "no factory found"